docs/design.md

```diff
@@ -0,0 +1,339 @@
+
+`Usage: trc-client [options] output <command id>`
+
+Stream the output of a command
+
+
+Output:
+```
+command output line1
+command output line2
+command output line3
+```
+### <a name="_xwvk9ga52s"></a>stop subcommand
+The stop command stops a running command. If the command is not running, or stopping the command fails, an error is returned.
+`Usage: trc-client [options] stop <command id>`
+Output:
+```json
+{
+     "success": "true"
+}
+```
+
+# <a name="_lzxdkro76353"></a>trc-agent usage
+
+The agent requires the following options:
+`Usage: trc-agent -ca-cert -key -host-cert`
+
+Options:
+
+  -ca-cert string
+	path to the host's self signed CA certificate
+
+  -key string
+	path to the hosts key file
+
+  -host-cert
+	path to the host cert signed by the ca-cert
+
+
+
+
+# <a name="_l5xtktkosihk"></a>gRPC Protocol
+The protobuf message definitions for sending and receiving commands between the client and agent:
+```
+service TrcAgent{
+     rpc Start(StartRequest) returns (StartResponse);
+     rpc Output(IdRequest) returns (stream OutputResponse);
+     rpc Status(IdRequest) returns (StatusResponse);
+     rpc Stop(IdRequest) returns (StopResponse);
+}
+
+message StartRequest {
+     string command = 1;
+     repeated string args = 2;
+}
+
+message StartResponse {
+     UUID id = 1;
+}
+
+message IdRequest {
+     UUID id = 1;
+}
+
+message OutputResponse {
+     bytes output = 1;
+}
+
+message StatusResponse {
+     UUID id = 1;
+     string cmd = 2;
+     repeated string args = 3;
+     string state = 4;
+     string error = 5;
+     int32 exit = 6;
+     Error error = 7;
+}
+
+message StopResponse {
+     UUID id = 1;
+}
+
+enum State {
+     RUNNING = 0;
+     COMPLETE = 1;
+     STOPPED = 2;
+     ERROR = 3;
+}
+
+message UUID {
+     string value = 1;
+}
+
+
+```
+
+# <a name="_lmohbtf2asqg"></a>Authentication and Authorization
+Authentication will be handled using mutual TLS (mTLS).  Both the agent and the client will generate self signed certs using a common CA. When the connection is established, the client and agent will each verify the other’s authenticity.   Each client user will generate their own cert. The agent can retrieve the cert from gRPC connection and, since each cert is unique to a single user, the agent can distinguish which user executes which commands.  
+
+Authorization will ensure that a client user (as determined by the cert attached to the request) can only get the stop, get the status and stream the output of commands started by that user.  In other words, if user 1 starts a command, user 2 can’t stop, status or get the output of that command. Internally this will be accomplished by pairing the command execution with the public certificate attached to the request.

```

Comment:
So, if a new cert is signed for the same user, they won't be able to stop, status and get output of previous commands.
I'm not sure that's desirable.
I think storing the user identifier would be better.
What do you think?